src/components/views/rooms/Autocomplete.js

```diff
@@ -1,36 +1,48 @@
 import React from 'react';
 import ReactCSSTransitionGroup from 'react-addons-css-transition-group';
 import classNames from 'classnames';
+import _ from 'lodash';
 
 import {getCompletions} from '../../../autocomplete/Autocompleter';
 
 export default class Autocomplete extends React.Component {
     constructor(props) {
         super(props);
+
+        this.onConfirm = this.onConfirm.bind(this);
+
         this.state = {
+            // list of completionResults, each containing completions
             completions: [],
 
+            // array of completions, so we can look up current selection by offset quickly
+            completionList: [],
+
             // how far down the completion list we are
-            selectionOffset: 0
+            selectionOffset: 0,
         };
     }
 
     componentWillReceiveProps(props, state) {
-        if(props.query == this.props.query) return;
+        if (props.query === this.props.query) {
+            return;
+        }
 
-        getCompletions(props.query, props.selection).map(completionResult => {
+        getCompletions(props.query, props.selection).forEach(completionResult => {
             try {
                 completionResult.completions.then(completions => {
                     let i = this.state.completions.findIndex(
                         completion => completion.provider === completionResult.provider
                     );
 
-                    i = i == -1 ? this.state.completions.length : i;
+                    i = i === -1 ? this.state.completions.length : i;
                     let newCompletions = Object.assign([], this.state.completions);
                     completionResult.completions = completions;
                     newCompletions[i] = completionResult;
+
                     this.setState({
-                        completions: newCompletions
+                        completions: newCompletions,
+                        completionList: _.flatMap(newCompletions, provider => provider.completions),
```

Comment:
Lodash has a bunch of useful utilities which can replace a decent amount of
code that's currently in the codebase. ratelimitedfunction, for example.

On Mon 4 Jul, 2016, 4:13 PM David Baker, <notifications@github.com> wrote:

> In src/components/views/rooms/Autocomplete.js
> <https://github.com/matrix-org/matrix-react-sdk/pull/296#discussion_r69440073>
> :
>
> >                      this.setState({
> > -                        completions: newCompletions
> > +                        completions: newCompletions,
> > +                        completionList: _.flatMap(newCompletions, provider => provider.completions),
>
> Is it worth pulling in all of lodash just for this?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/matrix-org/matrix-react-sdk/pull/296/files/a74db3a815879558110fcaa4b2f6f322dc1af783..cccc58b47f77f0eec644bc2455916496ed468318#r69440073>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAOlRHqTcDZbZTyo-RK80t7msI-bmPF_ks5qSOPBgaJpZM4Irc8R>
> .
>